help/sol.6
.\" Copyright (C) 2007 Martin Ferrari <martin.ferrari@gmail.com>
.\"
.\" This is free software; you may redistribute it and/or modify
.\" it under the terms of the GNU General Public License as
.\" published by the Free Software Foundation; either version 3,
.\" or (at your option) any later version.
.\"
.\" This is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\"You should have received a copy of the GNU General Public License along
.\"with this program; if not, write to the Free Software Foundation, Inc.,
.\"51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
.\"
.TH SOL 6 "March 22, 2007" "AisleRiot"
.\" Please adjust this date whenever revising the manpage.
.\"

.SH NAME
AisleRiot \- a collection of card games which are easy to play with the aid of a mouse.

.SH SYNOPSIS
.B sol
.RI [ options ]

.SH DESCRIPTION
This manual page documents briefly the
.B sol
command. For more information, please refer to the excellent GNOME help for
.B AisleRiot

.PP
.B AisleRiot
(also known as 
.BR Solitaire " or " sol )
is a collection of card games which are easy to play with the aid of a mouse.
The rules for the games have been coded for your pleasure in the GNOME
scripting language (Scheme).

The following games can be selected from within the program, or by using a command line switch:

.RS
Peek, Auld Lang Syne, Fortunes, Seahaven, King Albert, First Law, Straight Up,
Jumbo, Accordion, Ten Across, Plait, Lady Jane, Gypsy, Neighbor, Jamestown,
Osmosis, Kings Audience, Glenwood, Gay Gordons, Monte Carlo, Kansas, Camelot,
Fourteen, Scorpion, Isabel, Escalator, Agnes, Bristol, Quatorze, Bear River,
Gold Mine, Athena, Spiderette, Chessboard, Backbone, Yukon, Union Square,
Eight Off, Napoleons Tomb, Forty Thieves, Streets And Alleys, Maze, Clock,
Pileon, Canfield, Thirteen, Bakers Game, Triple Peaks, Easthaven, Terrace,
Aunt Mary, Carpet, Sir Tommy, Diamond Mine, Yield, Labyrinth, Thieves,
Saratoga, Cruel, Block Ten, Will O The Wisp, Odessa, Eagle Wing, Treize, Zebra,
Cover, Elevator, Fortress, Giant, Spider, Gaps, Bakers Dozen, Whitehead,
Freecell, Helsinki, Spider Three Decks, Scuffle, Poker, Klondike Three Decks,
Valentine, Royal East, Thumb And Pouch, Klondike, Doublets, Template, Golf,
Westhaven, Beleaguered Castle, Hopscotch, Eliminator, Aunt Mary,
Hamilton, Wall, Australian Patience
.RE

.SH OPTIONS
AisleRiot accepts the following options, besides the usual Gtk+/GNOME
switches:
.TP
.B \-v, \-\-variation=NAME
Select the game to play
.TP
.B \-\-display=DISPLAY
X display to use

.SH USAGE
This text is taken verbatim from the AisleRiot Manual V2.12.

Moving the cards is reasonably obvious. Assuming you have a right-handed mouse,
left-click on a card and drag it somewhere. Release the mouse button to deposit
the card.
 
If a card can be moved to the foundation(s), you do not need to drag it. You
can double-click on it and it will move. If a sequence of cards can be moved to
the foundations, you can generally move the sequence as one move.

Double clicking a card will send it to a foundation, if that is possible. This
is useful for cleaning up large numbers of cards at the end of a successful
game.

In games which have a stock pile you can redeal it by clicking on the empty
space where it was. Be warned that some games only let you do this a limited
number of time. Watch for the Redeals Left message at the bottom of the
window.
 
A useful tip to know is that right-clicking on a face-up card which is
partially obscured by a card laid over it will raise it so that you can see
what it is.
 
There is an alternative way to move cards. By selecting
.B Click to move
from the 
.B Control
menu you can click once to pick up a card and then click again to place it.
This can be faster than dragging and will be easier on your hand since you
don't have to hold the mouse down. However, it may take a while to get used to.
 
Some games have options to change the way the game is played. For example
Klondike can be played with either single card deals or three-card deals. In
these games there is a separate menu with the game's name. You can only change
the rules of the game at the beginning of the game -- the menu will be
inactive during the game.

Statistics on how well you have done at a particular game are available by
selecting
.B Statistics
from the Game menu. Only games where you actually start moving cards are
counted. The statistics are here for your own enjoyment, there are numerous
ways to cheat and it is unwise to make comparisons.

.SH HOMEPAGE
https://wiki.gnome.org/Apps/Aisleriot

.SH AUTHOR
This manual page was written by Martin Ferrari <martin.ferrari@gmail.com>,
for the Debian system (but may be used by others).
You can find the list of authors and more information about AisleRiot in
.IR /usr/share/doc/aisleriot .


NEWS
Aisleriot 3.18.0
================
* New games: Hamilton, Wall
* Translation updates

Aisleriot 3.12.0
================

* Build fixes (#715120, #726780)
* Games fixes (#677197, Vincent Povirk)
* Translation updates

Aisleriot 3.10.0
================
- Build fixes
- Translation updates

Aisleriot 3.9.90
================
- Many game improvements (Vincent Povirk)
- New card decks (Mario Frasca)
- Build fixes
- Translation updates

Aisleriot 3.6.2
===============
- translation updates

Aisleriot 3.6.1
===============
- translation updates

Aisleriot 3.6.0
===============
- fix help link
- translation updates

Aisleriot 3.5.90
================
- many games fixes (Vincent Povirk)
- translation updates

Aisleriot 3.5.0
===============
- build fixes
- many games fixes (Vincent Povirk)
- many translation updates

Aisleriot 3.3.2
===============
- Misc improvements and bug fixes
- Translation updates

Aisleriot 3.3.1
===============
- Aisleriot now requires guile >= 2.0
- Many improvements and bug fixes
- Translation updates

Aisleriot 3.3.0
===============
- Change card theme format
- Many game fixes (Vincent Povirk)
- Translation updates

Aisleriot 3.2.2
===============
- Dist missing sound files

Aisleriot 3.2.1
===============
- Minor bug fixes
- Translation updates

Aisleriot 3.2.0
===============
- Update card theme install info for debian (#659997, Josselin Mouette)
- Translation updates

Aisleriot 3.1.92
================
- Distribute a copy of the LGPL
- Update card theme install info for opensuse and ubuntu (#648601, Vincent Untz)
- Update card theme install info for ubuntu (#648601)
- Fix the build with the 'file' help method (#656967, Alexandre Rostovtsev)
- Build ghelp with yelp-tools instead of gnome-doc-utils (#657683, Jeremy Bicha)
- Fix game restart (#657735)
- Translation updates

Aisleriot 3.1.1
===============

* New game "Aunt Mary" (bug #325039)
* Translation updates

Aisleriot 3.1.0
===============

* Translation updates

Aisleriot 3.0.0
===============

Initial standalone release. Identical to Aisleriot from 
GNOME Games 3.0.0.



src/sol.c
/*
 * Copyright © 1998, 2001, 2003, 2006 Jonathan Blandford <jrb@alum.mit.edu>
 * Copyright © 2007 Christian Persch
 * Copyright © 2007 Andreas Røsdal <andreasr@gnome.org> 
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <config.h>

#include <string.h>

#include <libguile.h>

#include <glib.h>
#include <glib/gi18n.h>

#include <gtk/gtk.h>

#include "ar-debug.h"
#include "ar-stock.h"
#include "ar-runtime.h"
#include "ar-sound.h"

#include "ar-string-utils.h"
#include "conf.h"
#include "window.h"
#include "game.h"
#include "util.h"
#include "ar-application.h"

#if 0
/* String reserve */
N_("Solitaire")
N_("GNOME Solitaire")
N_("About Solitaire")
#endif /* 0 */

typedef struct {
  char *variation;
  gint seed; /* unused */
  gboolean freecell;
} AppData;

static void
add_main_options (GOptionContext *option_context,
                  AppData *data)
{
  const GOptionEntry aisleriot_options[] = {
    { "variation", 'v', 0, G_OPTION_ARG_STRING, &data->variation,
      N_("Select the game type to play"), N_("NAME") },
    { "freecell", 0, G_OPTION_FLAG_HIDDEN, G_OPTION_ARG_NONE, &data->freecell,
      NULL, NULL },

    /* Ignored option, for backward compat with saved session */
    { "seed", 's', G_OPTION_FLAG_HIDDEN, G_OPTION_ARG_STRING, &data->seed,
      NULL, NULL },

    { NULL }
  };

  g_option_context_add_main_entries (option_context,
				     aisleriot_options, GETTEXT_PACKAGE);
}

static void
main_prog (void *closure, int argc, char *argv[])
{
  AppData data;
  GOptionContext *option_context;
  GError *error = NULL;
  gboolean retval;
  GtkApplication *application;
  int status G_GNUC_UNUSED;

  memset (&data, 0, sizeof (AppData));

  option_context = g_option_context_new (NULL);
  g_option_context_set_translation_domain (option_context, GETTEXT_PACKAGE);

  add_main_options (option_context, &data);

  g_option_context_add_group (option_context, gtk_get_option_group (TRUE));

  retval = g_option_context_parse (option_context, &argc, &argv, &error);
  g_option_context_free (option_context);

  if (!retval) {
    g_printerr ("%s\n", error->message);
    g_error_free (error);
    goto cleanup;
  }

  g_set_application_name (data.freecell ? _("FreeCell Solitaire") : _("AisleRiot"));

  aisleriot_conf_init ();

  /* If we are asked for a specific game, check that it is valid. */
  if (!data.freecell &&
      data.variation != NULL) {
    char *game_module = NULL;

    if (data.variation[0] != '\0') {
      game_module = ar_filename_to_game_module (data.variation);
    }

    g_free (data.variation);
    data.variation = game_module;
  }

  if (!data.freecell && !data.variation) {
    char *pref;

    pref = ar_conf_get_string_with_default (NULL, aisleriot_conf_get_key (CONF_VARIATION), DEFAULT_VARIATION);
    data.variation = ar_filename_to_game_module (pref);
    g_free (pref);
  }

  g_assert (data.variation != NULL || data.freecell);

  application = ar_application_new (data.variation, data.freecell);
  status = g_application_run (G_APPLICATION (application), 0, NULL);
  g_object_unref (application);

  aisleriot_conf_shutdown ();

cleanup:
  g_free (data.variation);

  ar_runtime_shutdown ();
}

int
main (int argc, char *argv[])
{
#if !GTK_CHECK_VERSION (3, 19, 5)
  if (!gtk_check_version(3, 19, 5)) {
    g_printerr("Aisleriot needs to be recompiled against gtk+ >= 3.19.5 in order to run with gtk+ >= 3.19.5\n");
    return 1;
  }
#endif

  if (!ar_runtime_init ("aisleriot"))
    return 1;

  g_setenv ("GUILE_WARN_DEPRECATED", "detailed", TRUE);
  g_setenv ("GUILE_AUTO_COMPILE", "0", TRUE);

  /* Set some env vars to disable the ubuntu modules. They'll certainly patch this
   * out in their package, but anyone running from git will get the right
   * behaviour.
   */
  g_setenv ("LIBOVERLAY_SCROLLBAR", "0", TRUE);
  g_setenv ("UBUNTU_MENUPROXY", "0", TRUE);
  g_setenv ("NO_UNITY_GTK_MODULE", "1", TRUE);

  /* Not interested in silly debug spew polluting the journal, bug #749195 */
  if (g_getenv ("G_ENABLE_DIAGNOSTIC") == NULL)
    g_setenv ("G_ENABLE_DIAGNOSTIC", "0", TRUE);

  scm_boot_guile (argc, argv, main_prog, NULL); /* no return */

  return 0;
}
